app/AccountComponent.scala

```diff
@@ -1,22 +1,51 @@
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.ActorContext
 import play.api.mvc.ControllerComponents
 import akka.cluster.sharding.typed.scaladsl._
 import play.scaladsl.cqrs._
 import com.softwaremill.macwire._
 import model._
-import _root_.controllers.AccountController
+import controllers.AccountController
 
 trait AccountComponent extends ClusterShardingComponents with CqrsComponents {
 
   lazy val accountController = wire[AccountController]
 
   def controllerComponents: ControllerComponents
 
-  // register behavior as a sharded entity
-  clusterSharding.init(
-    Entity(
-      Account.typeKey,
-      ctx => Account.behavior(ctx)
-    )
+  // OPTION A:
+  // This is the regular usage of the factory.
+  //  - CqrsComponents#createEntityFactory isolates the user from the sharding
+  //    aspects so the `behaviorFRunc` argument takes (ActorContext[], PersistenceId)
+  createEntityFactory(
+    Account.typeKey.name,
+    Account.behavior,
+    Tagger[AccountEvent].addTagGroup("TagA", 3)
   )
 
+  // OPTION B:
+  // This is how you use it when you want an extra ActorContext on your EventSourcedBehavior
+  //  - ignore the `createEntityFactory()` method
+  //  - create an anonymous sub-class of EntityFactory and override `defaultSharder`
+  new EntityFactory(
+    Account.typeKey.name + "-on-steroids",
+    Account.behavior,
+    Tagger[AccountEvent].addTagGroup("TagA", 3),
+    actorSystem
+  ) {
+
+    // the sharder is what I call the function that a ShardRegion will use to spawn
+    // a new Behavior on the receiving node. It gets an actorContext and a entityContext
+    // because it is used directly from the ShardRegion.
+    override def defaultSharder(
+        actorContext: ActorContext[AccountCommand[_]],
+        entityContext: EntityContext
+    ): Behavior[AccountCommand[_]] = {
+      val spr = super.defaultSharder _
+      actorContext.log.info(s"Account: ${entityContext.entityId} instatiated")
+      spr(actorContext, entityContext)
+    }
+
+  }
```

Comment:
This whole OPTION A vs OPTION B is the important part of this PR.

Here, OPTION A represents the common usage of the `EntityFactory` where users _just_ take the `CqrsComponents#createEntityFactory` and move on.

Option B, OTOH, is for users who don't deal with all the Akka setup but need access to an `ActorContext` or maybe a little something more. I've just done the minimal to port the original logging example to here so I'm not 100% sure this is a great API for a general purpose (where users can do anything they like, safely, with the provided `ActorContext`.

Reply:
Having a method where you pass the ActorContext doesn't solve the problem because `setup` is not the only wrapper that a user may need. 

There are many more: 

* Behaviors.monitor()
* Behaviors.supervise()
* Behaviors.intercept()
* Behaviors.withTimers()

To name a few.

Reply:
Fair enough. Just use:

```scala

    // the sharder is what I call the function that a ShardRegion will use to spawn
    // a new Behavior on the receiving node. It gets an actorContext and a entityContext
    // because it is used directly from the ShardRegion.
    override def defaultSharder(
        actorContext: ActorContext[AccountCommand[_]],
        entityContext: EntityContext
    ): Behavior[AccountCommand[_]] = {
      val spr = super.defaultSharder _
      actorContext.log.info(s"Account: ${entityContext.entityId} instatiated")
      spr(actorContext, entityContext)
    }
```

And use the return value of `super.defaultSharder` as an argument on any of the factory methods you mentioned.

We could even extract the `Behaviors.setup` from `EntityFactory#defaultSharder`, I think. 

I think that by implementing `buildEntity` as:

```scala
  // is concerned about sharding, closes over an `entityTypeKey`and 
  // produces an `Entity`
  private def buildEntity(): Entity[Command, ShardingEnvelope[Command]] = {
    Entity(
      typeKey,
      entityContext => ... // use defaultSharder here, somehow
    )
  }

  // This is the glue code connecting sharding and persistence. But we still
  // leave the door open to producing regular `Behavior` so users can 
  // override and wrap this.
  protected def defaultSharder(
         ... some extra args??? ...,
        entityContext: EntityContext
       ): Behavior[Command] = {
    val peid: PersistenceId = asPersistenceId(entityContext.entityId)
    behaviorFunc(peid).withTagger(...)
  }

  // where `behaviorFunc: PersistenceId => EventSourcedBehavior`
```

where `defaultSharder` is protected, we open the door to wrapping `behaviorFunc: PersistenceId => EventSourcedBehavior` in whichever factory method from `Behaviors` the user wanted.

Reply:
We can, at most, dedicate an extra hour pursuing this idea, but after all the time we've already spent I'm getting more and more convinced that providing an extra abstraction is not necessary and we may be better off putting these effort aside for a while.

Reply:
I'll close this Pr for the time being. Feel free to reopen if you think this is still worth a final round.